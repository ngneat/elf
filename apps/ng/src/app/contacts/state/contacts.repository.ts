import { Injectable } from '@angular/core';
import { createState, Store } from '@ngneat/elf';

import {
  PaginationData,
  selectCurrentPage,
  selectCurrentPageEntities,
  setCurrentPage,
  setPage,
  updatePaginationData,
  withPagination,
} from '@ngneat/elf-pagination';
import { addEntities, withEntities } from '@ngneat/elf-entities';

interface Contact {
  id: number;
  email: string;
  name: string;
  address: string;
}

const { state, config } = createState(
  withPagination(),
  withEntities<Contact>()
);

const store = new Store({ state, config, name: 'contacts' });

@Injectable({ providedIn: 'root' })
export class ContactsRepository {
  activePage$ = store.pipe(selectCurrentPage());
  activePageContacts$ = store.pipe(selectCurrentPageEntities());

  setActivePage(id: Contact['id']) {
    store.update(setCurrentPage(id));
  }

  addContacts(page: number, contacts: PaginationData & { data: Contact[] }) {
    const { data, ...pagination } = contacts;

    store.update(
      addEntities(data),
      updatePaginationData(pagination),
      setPage(
        page,
        data.map((c) => c.id)
      )
    );
  }

  get store() {
    return store;
  }
}
