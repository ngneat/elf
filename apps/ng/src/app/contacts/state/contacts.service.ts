import { Injectable } from '@angular/core';
import { getData } from '../contacts-data';
import { timer } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { ContactsRepository } from './contacts.repository';
import { skipWhilePageExists } from '@ngneat/elf-pagination';

@Injectable({ providedIn: 'root' })
export class ContactsService {
  constructor(private repo: ContactsRepository) {}

  getContacts(page: number) {
    return timer(300).pipe(
      map(() => getData({ page })),
      tap((res) => this.repo.addContacts(page, res)),
      skipWhilePageExists(this.repo.store, page)
    );
  }
}
