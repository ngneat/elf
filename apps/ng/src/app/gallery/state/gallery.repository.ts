import { Injectable } from '@angular/core';
import { createStore } from '@ngneat/elf';
import {
  addEntities,
  selectAllEntities,
  selectEntities,
  UIEntitiesRef,
  unionEntities,
  updateEntities,
  withEntities,
  withUIEntities,
} from '@ngneat/elf-entities';

export interface GalleryItem {
  id: number;
  title: string;
  description: string;
  image: string;
}

export interface GalleryUIItem {
  id: number;
  open: boolean;
}

const store = createStore(
  { name: 'gallery' },
  withEntities<GalleryItem>(),
  withUIEntities<GalleryUIItem>()
);

@Injectable({ providedIn: 'root' })
export class GalleryRepository {
  items$ = this.store
    .combine({
      entities: store.pipe(selectAllEntities()),
      UIEntities: store.pipe(selectEntities({ ref: UIEntitiesRef })),
    })
    .pipe(unionEntities());

  addGalleryItems(galleryItems: GalleryItem[]) {
    const uiGalleryItems = galleryItems.map(({ id }) => ({ id, open: false }));
    store.update(
      addEntities(galleryItems),
      addEntities(uiGalleryItems, { ref: UIEntitiesRef })
    );
  }

  get store() {
    return store;
  }

  toggleItemOpen(galleryItemId: number) {
    this.store.update(
      updateEntities(galleryItemId, (item) => ({ ...item, open: !item.open }), {
        ref: UIEntitiesRef,
      })
    );
  }
}
