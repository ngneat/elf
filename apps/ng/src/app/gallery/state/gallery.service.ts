import { Injectable } from '@angular/core';
import { getData } from '../gallery-data';
import { timer } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { GalleryRepository } from './gallery.repository';

@Injectable({ providedIn: 'root' })
export class GalleryService {
  constructor(private repo: GalleryRepository) {}

  getGalleryItems() {
    return timer(300).pipe(
      map(() => getData()),
      tap((res) => this.repo.addGalleryItems(res))
    );
  }
}
