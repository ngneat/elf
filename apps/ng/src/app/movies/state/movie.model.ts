export type Movie = {
  id: number | string;
  title: string;
  genres: (string | number)[];
  actors: (string | number)[];
};

export interface FullMovie {
  id: number | string;
  title: string;
  genres: { id: number | string; name: string }[];
  actors: { id: number | string; name: string }[];
}
