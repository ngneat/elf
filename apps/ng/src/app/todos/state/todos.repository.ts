import { Injectable } from '@angular/core';
import { switchMap } from 'rxjs/operators';
import { write } from '../../store/mutations';
import { createState, select, Store, withProps } from '@ngneat/elf';
import {
  addEntities,
  selectAll,
  selectAllApply,
  updateEntities,
  withEntities,
} from '@ngneat/elf-entities';

interface Todo {
  id: number;
  title: string;
  completed: boolean;
}

export interface TodosProps {
  filter: 'ALL' | 'ACTIVE' | 'COMPLETED';
}

const { state, config } = createState(
  withEntities<Todo>(),
  withProps<TodosProps>({ filter: 'ALL' })
);

const store = new Store({ name: 'todos', state, config });

@Injectable({ providedIn: 'root' })
export class TodosRepository {
  todos$ = store.pipe(selectAll());
  filter$ = store.pipe(select((state) => state.filter));

  visibleTodos$ = this.filter$.pipe(
    switchMap((filter) => {
      return store.pipe(
        selectAllApply({
          filterEntity({ completed }) {
            if (filter === 'ALL') return true;
            return filter === 'COMPLETED' ? completed : !completed;
          },
        })
      );
    })
  );

  addTodo(title: Todo['title']) {
    store.update(addEntities({ id: Math.random(), title, completed: false }));
  }

  updateFilter(filter: TodosProps['filter']) {
    store.update(
      write((state) => {
        state.filter = filter;
      })
    );
  }

  updateCompleted(id: Todo['id']) {
    store.update(
      updateEntities(id, (entity) => ({
        ...entity,
        completed: !entity.completed,
      }))
    );
  }
}
