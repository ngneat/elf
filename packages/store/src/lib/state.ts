type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;
type Merge<State extends any[], Key extends PropertyKey> = UnionToIntersection<
  State[number][Key]
>;

export type PropsFactory<Props, Config> = { props: Props; config: Config };
export type EmptyConfig = undefined;

export function createState<S extends PropsFactory<any, any>[]>(
  ...propsFactories: S
): { state: Merge<S, 'props'>; config: Merge<S, 'config'> } {
  const result = { config: {}, state: {} };

  for (const { config, props } of propsFactories) {
    Object.assign(result.config, config);
    Object.assign(result.state, props);
  }

  return result as { state: Merge<S, 'props'>; config: Merge<S, 'config'> };
}
