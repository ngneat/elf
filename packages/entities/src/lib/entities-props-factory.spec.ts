import { createStore } from '@ngneat/elf';
import { expectTypeOf } from 'expect-type';
import { selectAllEntities, selectEntities, withEntities } from '..';
import { addEntities } from './add.mutation';
import { entitiesPropsFactory } from './entity.state';

const { cartEntitiesRef, withCartEntities } = entitiesPropsFactory('cart');

describe('entities props factory', () => {
  it('should create entities', () => {
    const store = createStore(
      { name: 'todos' },
      withCartEntities<{ title: string; id: number }>()
    );

    store.update(
      addEntities({ id: 1, title: 'foo' }, { ref: cartEntitiesRef })
    );

    expect(store.getValue()).toMatchSnapshot();
  });

  it('should work with the default entities', () => {
    const store = createStore(
      { name: 'todos' },
      withEntities<{ id: string; label: string }>(),
      withCartEntities<{ title: string; id: number }>()
    );

    store.update(
      addEntities({ id: '1', label: 'foo' }),
      addEntities({ id: 1, title: 'foo' }, { ref: cartEntitiesRef })
    );

    expect(store.getValue()).toMatchSnapshot();

    expectTypeOf(store.getValue()).toEqualTypeOf<{
      entities: Record<
        string,
        {
          id: string;
          label: string;
        }
      >;
      ids: string[];
      cartEntities: Record<
        number,
        {
          title: string;
          id: number;
        }
      >;
      cartIds: number[];
    }>();
  });

  it('should infer types', () => {
    const store = createStore(
      { name: 'todos' },
      withCartEntities<{ title: string; id: number }>()
    );

    expectTypeOf(store.getValue()).toEqualTypeOf<{
      cartEntities: Record<
        number,
        {
          title: string;
          id: number;
        }
      >;
      cartIds: number[];
    }>();

    try {
      store.update(
        // @ts-expect-error - The default entities isn't declared
        addEntities({ id: '1', label: 'foo' }),
        addEntities({ id: 2, title: 'foo' }, { ref: cartEntitiesRef })
      );
    } catch {
      //
    }

    store.update(
      // @ts-expect-error - id should be a number
      addEntities({ id: '1', title: 'foo' }, { ref: cartEntitiesRef }),
      // @ts-expect-error - nope isn't exists on type of cart entity
      addEntities({ id: 2, title: 'foo', nope: '' }, { ref: cartEntitiesRef })
    );
  });

  it('should work with multiple', () => {
    interface Actor {
      id: string;
      name: string;
    }

    interface Genre {
      id: string;
      name: string;
    }

    interface Movie {
      id: string;
      title: string;
      genres: Array<Genre['id']>;
      actors: Array<Actor['id']>;
    }

    const { actorsEntitiesRef, withActorsEntities } =
      entitiesPropsFactory('actors');
    const { genresEntitiesRef, withGenresEntities } =
      entitiesPropsFactory('genres');

    const store = createStore(
      { name: 'movies' },
      withEntities<Movie>(),
      withGenresEntities<Genre>(),
      withActorsEntities<Actor>()
    );

    expect(store.getValue()).toMatchSnapshot();

    const spy = jest.fn();

    store
      .combine({
        movies: store.pipe(selectAllEntities()),
        generes: store.pipe(selectEntities({ ref: genresEntitiesRef })),
        actors: store.pipe(selectEntities({ ref: actorsEntitiesRef })),
      })
      .subscribe((v) => {
        spy(v);

        expectTypeOf(v).toEqualTypeOf<{
          movies: Movie[];
          generes: Record<string, Genre>;
          actors: Record<string, Actor>;
        }>();
      });

    expect(spy).toHaveBeenCalledTimes(1);

    store.update(
      addEntities({ id: '1', name: 'foo' }, { ref: actorsEntitiesRef }),
      addEntities({ id: '1', name: 'foo' }, { ref: genresEntitiesRef }),
      addEntities({ id: '1', title: 'one', genres: ['1'], actors: ['1'] })
    );

    expect(store.getValue()).toMatchSnapshot();

    expect(spy).toHaveBeenCalledTimes(2);
  });

  it('should allow initial value', () => {
    const { withFooEntities, fooEntitiesRef } = entitiesPropsFactory('foo');

    const store = createStore(
      { name: 'foo' },
      withFooEntities<{ id: number; count: number }>({
        initialValue: [
          {
            id: 1,
            count: 0,
          },
        ],
      })
    );

    expect(store.getValue()).toMatchInlineSnapshot(`
      Object {
        "fooEntities": Object {
          "1": Object {
            "count": 0,
            "id": 1,
          },
        },
        "fooIds": Array [
          1,
        ],
      }
    `);

    store.update(addEntities({ id: 2, count: 1 }, { ref: fooEntitiesRef }));

    expect(store.getValue()).toMatchInlineSnapshot(`
      Object {
        "fooEntities": Object {
          "1": Object {
            "count": 0,
            "id": 1,
          },
          "2": Object {
            "count": 1,
            "id": 2,
          },
        },
        "fooIds": Array [
          1,
          2,
        ],
      }
    `);
  });
});
