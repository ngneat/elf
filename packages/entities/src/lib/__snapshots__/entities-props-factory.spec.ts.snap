// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entities props factory should create entities 1`] = `
Object {
  "cartEntities": Object {
    "1": Object {
      "id": 1,
      "title": "foo",
    },
  },
  "cartIds": Array [
    1,
  ],
}
`;

exports[`entities props factory should work with multiple 1`] = `
Object {
  "actorsEntities": Object {},
  "actorsIds": Array [],
  "entities": Object {},
  "genresEntities": Object {},
  "genresIds": Array [],
  "ids": Array [],
}
`;

exports[`entities props factory should work with multiple 2`] = `
Object {
  "actorsEntities": Object {
    "1": Object {
      "id": "1",
      "name": "foo",
    },
  },
  "actorsIds": Array [
    "1",
  ],
  "entities": Object {
    "1": Object {
      "actors": Array [
        "1",
      ],
      "genres": Array [
        "1",
      ],
      "id": "1",
      "title": "one",
    },
  },
  "genresEntities": Object {
    "1": Object {
      "id": "1",
      "name": "foo",
    },
  },
  "genresIds": Array [
    "1",
  ],
  "ids": Array [
    "1",
  ],
}
`;

exports[`entities props factory should work with the default entities 1`] = `
Object {
  "cartEntities": Object {
    "1": Object {
      "id": 1,
      "title": "foo",
    },
  },
  "cartIds": Array [
    1,
  ],
  "entities": Object {
    "1": Object {
      "id": "1",
      "label": "foo",
    },
  },
  "ids": Array [
    "1",
  ],
}
`;
